src/V3Timing.cpp

```diff
@@ -0,0 +1,785 @@
+            if (nodep->bodysp()) whilep->addNext(nodep->bodysp()->unlinkFrBackWithNext());
+            nodep->replaceWith(whilep);
+        }
+        nodep->deleteTree();
+    }
+    virtual void visit(AstFork* nodep) override {
+        if (!nodep->joinType().joinNone()) {
+            v3Global.timing(true);
+            m_procp->user2(true);
+        }
+        VL_RESTORER(m_procp);
+        VL_RESTORER(m_underFork);
+        m_procp = nodep;
+        m_underFork = true;
+        iterateChildren(nodep);
+    }
+    virtual void visit(AstBegin* nodep) override {
+        VL_RESTORER(m_underFork);
+        m_underFork = false;
+        iterateChildren(nodep);
+    }
+
+    //--------------------
+    virtual void visit(AstNode* nodep) override { iterateChildren(nodep); }
+
+public:
+    // CONSTRUCTORS
+    explicit TimingVisitor(AstNetlist* nodep)
+        : m_timeprecision(nodep->timeprecision()) {
+        iterate(nodep);
+        if (debug() >= 6) m_depGraph.dumpDotFilePrefixed("timing_deps");
+    }
+    virtual ~TimingVisitor() override = default;
+};
+
+//######################################################################
+// Transform timing constructs to C++ statements
+
+class TimingToCStmtsVisitor final : public VNVisitor {
+private:
+    // NODE STATE
+    //  AstSenTree::user1()      -> AstVarScope*.  Trigger scheduler assigned to this sentree
+    //  AstSenTree::user2()      -> AstComment*.   Comment to put before this sentree's
+    //                                             corresponding await
+    //  AstCFunc::user1()        -> bool.  Set true if CFunc has timing awaits
+    const VNUser1InUse m_inuser1;
+    const VNUser2InUse m_inuser2;
+
+    // STATE
+    // Current context
+    AstScope* const m_scopeTopp;  // Scope at the top
+    AstScope* m_scopep = nullptr;  // Current scope
+    AstCFunc* m_cfuncp = nullptr;  // Current CFunc
+    AstSenTree* m_lastTimingSensp = nullptr;  // Last encountered EventControl sentree
+
+    // Unique names
+    V3UniqueNames m_forkNames{"__Vfork__"};  // Fork name generator
+    V3UniqueNames m_trigSchedNames{"__VtrigSched"};  // Trigger scheduler name generator
+
+    // DTypes
+    AstBasicDType* const m_forkDtp = new AstBasicDType{
+        m_scopeTopp->fileline(), VBasicDTypeKwd::FORK, VSigning::UNSIGNED};  // Fork variable type
+    AstBasicDType* const m_dlySchedDtp
+        = new AstBasicDType{m_scopeTopp->fileline(), VBasicDTypeKwd::DELAY_SCHEDULER,
+                            VSigning::UNSIGNED};  // Delay scheduler type
+    AstBasicDType* const m_trigSchedDtp
+        = new AstBasicDType{m_scopeTopp->fileline(), VBasicDTypeKwd::TRIGGER_SCHEDULER,
+                            VSigning::UNSIGNED};  // Trigger scheduler type
+
+    // Timing-related globals
+    AstVarScope* const m_dlySchedp
+        = m_scopeTopp->createTemp("__VdlySched", m_dlySchedDtp);  // Global delay scheduler
```

Comment:
Nit: Should create these lazily at some point, but you don't need to worry about it right now if you don't want to.

Reply:
Done